### PR TITLE
[Enhancement] Optimize aggregation from agg materialized view

### LINKED ISSUE #10324: aggregation above agg table can be optimized
## Enhancement

for table 
CREATE TABLE `table1` (
  `c1` datetime NULL COMMENT "",
  `c2` varchar(65533) NULL COMMENT "",
  `c3` varchar(65533) NULL COMMENT "",
  `c4` varchar(65533) NULL COMMENT "",
  `c5` varchar(65533) NULL COMMENT ""
) ENGINE=OLAP
DUPLICATE KEY(`c1`, `c2`, `c3`)
COMMENT "OLAP"
PARTITION BY RANGE(`c1`)
(PARTITION p20211001 VALUES [('2021-10-01 00:00:00'), ('2021-10-02 00:00:00')),
PARTITION p20211002 VALUES [('2021-10-02 00:00:00'), ('2021-10-03 00:00:00')))
DISTRIBUTED BY HASH(`c2`) BUCKETS 3
PROPERTIES (
"replication_num" = "1",
"in_memory" = "false",
"storage_format" = "DEFAULT",
"enable_persistent_index" = "false"
);

create mv:
create materialized view as
select c1,c2,c3 from table1 group by c1,c2,c3;

for query: select count(*) from (select c1,c2,c3 from table1 group by c1,c2,c3) t, the plan is as following:

| PLAN FRAGMENT 0                                                                 |
|  OUTPUT EXPRS:6: count                                                          |
|   PARTITION: UNPARTITIONED                                                      |
|                                                                                 |
|   RESULT SINK                                                                   |
|                                                                                 |
|   7:AGGREGATE (merge finalize)                                                  |
|   |  output: count(6: count)                                                    |
|   |  group by:                                                                  |
|   |                                                                             |
|   6:EXCHANGE                                                                    |
|                                                                                 |
| PLAN FRAGMENT 1                                                                 |
|  OUTPUT EXPRS:                                                                  |
|   PARTITION: HASH_PARTITIONED: 1: dteventtime, 2: vgameappid, 4: vopenid        |
|                                                                                 |
|   STREAM DATA SINK                                                              |
|     EXCHANGE ID: 06                                                             |
|     UNPARTITIONED                                                               |
|                                                                                 |
|   5:AGGREGATE (update serialize)                                                |
|   |  output: count(*)                                                           |
|   |  group by:                                                                  |
|   |                                                                             |
|   4:Project                                                                     |
|   |  <slot 1> : 1: dteventtime                                                  |
|   |                                                                             |
|   3:AGGREGATE (merge finalize)                                                  |
|   |  group by: 1: dteventtime, 2: vgameappid, 4: vopenid                        |
|   |                                                                             |
|   2:EXCHANGE                                                                    |
|                                                                                 |
| PLAN FRAGMENT 2                                                                 |
|  OUTPUT EXPRS:                                                                  |
|   PARTITION: RANDOM                                                             |
|                                                                                 |
|   STREAM DATA SINK                                                              |
|     EXCHANGE ID: 02                                                             |
|     HASH_PARTITIONED: 1: dteventtime, 2: vgameappid, 4: vopenid                 |
                                                                              |
|   1:AGGREGATE (update serialize)                                                |
|   |  STREAMING                                                                  |
|   |  group by: 1: dteventtime, 2: vgameappid, 4: vopenid                        |
|   |                                                                             |
|   0:OlapScanNode                                                                |
|      TABLE: table1                                                              |
|      PREAGGREGATION: ON                                                         |
|      partitions=100/131                                                         |
|      rollup: table1                                                             |
|      tabletRatio=300/300                                                        |
|      tabletList=11935,11937,11939,11905,11907,11909,11797,11799,11801,11809 ... |
|      cardinality=285214848                                                      |
|      avgRowSize=3.0                                                             |
|      numNodes=0

the plan can be optimized, and the group by agg can be removed because the mv has the partition and distribution keys

### PR DESCRIPTION
**Pull request by ABingHuang**
## What type of PR is this：
- [ ] bug
- [ ] feature
- [x] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes #10324

## Problem Summary(Required) ：

For aggregation on distribution and partition key of agg materialized view,
aggregation can be removed to optimize compute. In my test, the time cost
decrease from 7.21s to 1.21s for one hundred million records table.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] I have added user document for my new feature or new function

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 50 / 50 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleSet.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleSetType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/RemoveAggregationFromAggTable.java|42|42|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/PreAggregateTurnOnRule.java|4|4|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 51 / 51 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleSet.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleSetType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/RemoveAggregationFromAggTable.java|43|43|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/PreAggregateTurnOnRule.java|4|4|100.00%|[]|

File: fe/fe-core/src/test/java/com/starrocks/sql/plan/PlanFragmentWithCostTest.java [hunk 1]
```diff
@@ -1192,4 +1192,65 @@ public void testExpressionWithCTE() throws Exception {
...
+                "  PARTITION: RANDOM\n" +
+                "\n" +
+                "  STREAM DATA SINK\n" +
+                "    EXCHANGE ID: 02\n" +
+                "    UNPARTITIONED\n" +
+                "\n" +
+                "  1:Project\n" +
+                "  |  <slot 1> : 1: LO_ORDERDATE\n" +
+                "  |  <slot 2> : 2: LO_ORDERKEY\n" +
+                "  |  <slot 40> : sum(13: LO_REVENUE)\n" +
```
Comment (kangkaisen) [inline fe/fe-core/src/test/java/com/starrocks/sql/plan/PlanFragmentWithCostTest.java:1247]: Project operator couldn't do sum function

Comment (ABingHuang) [reply fe/fe-core/src/test/java/com/starrocks/sql/plan/PlanFragmentWithCostTest.java:1247]: fixed

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 53 / 53 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleSet.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleSetType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/RemoveAggregationFromAggTable.java|45|45|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/PreAggregateTurnOnRule.java|4|4|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 61 / 61 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleSet.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleSetType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/RemoveAggregationFromAggTable.java|53|53|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/PreAggregateTurnOnRule.java|4|4|100.00%|[]|

**Comment by ABingHuang** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 60 / 61 (98.36%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/PreAggregateTurnOnRule.java|4|5|80.00%|[286]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleSetType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/RemoveAggregationFromAggTable.java|53|53|100.00%|[]|

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RemoveAggregationFromAggTable.java [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        super(RuleType.TF_REMOVE_AGGREGATION_BY_AGG_TABLE,
+                Pattern.create(OperatorType.LOGICAL_AGGR).addChildren(Pattern.create(
+                        OperatorType.LOGICAL_OLAP_SCAN)));
+    }
+
+    @Override
+    public boolean check(OptExpression input, OptimizerContext context) {
+        OptExpression scanExpression = input.getInputs().get(0);
+        Preconditions.checkState(scanExpression.getOp() instanceof LogicalOlapScanOperator);
+        LogicalOlapScanOperator scanOperator = (LogicalOlapScanOperator) scanExpression.getOp();
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RemoveAggregationFromAggTable.java:49]: please consider group by sub_string(column) case.

Comment (ABingHuang) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RemoveAggregationFromAggTable.java:51]: limit it

**Comment by ABingHuang** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 62 / 63 (98.41%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/PreAggregateTurnOnRule.java|4|5|80.00%|[286]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleSetType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/RemoveAggregationFromAggTable.java|55|55|100.00%|[]|

**Review by kangkaisen — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java [hunk 3]
```diff
@@ -289,6 +290,7 @@ void logicalRuleRewrite(Memo memo, TaskContext rootTaskContext) {
         ruleRewriteIterative(memo, rootTaskContext, new MergeProjectWithChildRule());
         ruleRewriteOnlyOnce(memo, rootTaskContext, new GroupByCountDistinctRewriteRule());
         ruleRewriteOnlyOnce(memo, rootTaskContext, new ReorderIntersectRule());
+        ruleRewriteIterative(memo, rootTaskContext, new RemoveAggregationFromAggTable());
```
Comment (LiShuMing) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java:293]: Maybe we can add `RuleSetType` Of MVRewrite?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RemoveAggregationFromAggTable.java [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+import com.starrocks.sql.optimizer.rewrite.ReplaceColumnRefRewriter;
+import com.starrocks.sql.optimizer.rule.RuleType;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class RemoveAggregationFromAggTable extends TransformationRule {
+    private static List<String> unsupportedFunctionNames =
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RemoveAggregationFromAggTable.java:34]: static final

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RemoveAggregationFromAggTable.java [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+import com.starrocks.sql.optimizer.operator.scalar.ScalarOperator;
+import com.starrocks.sql.optimizer.rewrite.ReplaceColumnRefRewriter;
+import com.starrocks.sql.optimizer.rule.RuleType;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class RemoveAggregationFromAggTable extends TransformationRule {
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RemoveAggregationFromAggTable.java:33]: add comment

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RemoveAggregationFromAggTable.java [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+            partitionColumnNames.addAll(rangePartitionInfo.getPartitionColumns().stream()
+                    .map(column -> column.getName().toLowerCase()).collect(Collectors.toList()));
+        }
+
+        List<String> distributionColumnNames = olapTable.getDistributionColumnNames().stream()
+                .map(String::toLowerCase).collect(Collectors.toList());
+
+        LogicalAggregationOperator aggregationOperator = (LogicalAggregationOperator) input.getOp();
+        for (Map.Entry<ColumnRefOperator, CallOperator> entry : aggregationOperator.getAggregations().entrySet()) {
+            if (unsupportedFunctionNames.contains(entry.getValue().getFnName().toLowerCase())) {
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RemoveAggregationFromAggTable.java:83]: need check the input of aggregate function is materialized column?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RemoveAggregationFromAggTable.java [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+    @Override
+    public List<OptExpression> transform(OptExpression input, OptimizerContext context) {
+        LogicalAggregationOperator aggregationOperator = (LogicalAggregationOperator) input.getOp();
+        Map<ColumnRefOperator, ScalarOperator> projectMap = Maps.newHashMap();
+        aggregationOperator.getGroupingKeys().forEach(g -> projectMap.put(g, g));
+        for (Map.Entry<ColumnRefOperator, CallOperator> entry : aggregationOperator.getAggregations().entrySet()) {
+            // in this case, CallOperator must have only one child ColumnRefOperator
+            projectMap.put(entry.getKey(), entry.getValue().getChild(0));
+        }
+        if (aggregationOperator.getProjection() != null) {
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RemoveAggregationFromAggTable.java:103]: call this rule after MaterializedViewRewrite, and you don't need handle projection

**Comment by Seaven** [comment]
Why MVRewrite wasn't hand the case?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java (hunk 3 above)
Comment (ABingHuang) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java:293]: it seems not necessary to add a RuleSetType

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RemoveAggregationFromAggTable.java (hunk 7 above)
Comment (ABingHuang) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RemoveAggregationFromAggTable.java:130]: I test this way, it is not ok.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RemoveAggregationFromAggTable.java (hunk 4 above)
Comment (ABingHuang) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RemoveAggregationFromAggTable.java]: done

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RemoveAggregationFromAggTable.java (hunk 6 above)
Comment (ABingHuang) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RemoveAggregationFromAggTable.java:92]: done

**Review by Seaven — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 76 / 80 (95.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/PreAggregateTurnOnRule.java|4|5|80.00%|[286]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/RemoveAggregationFromAggTable.java|70|73|95.89%|[55, 100, 106]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|[]|

**Comment by ABingHuang** [comment]
run starrocks_admit_test

**Review by Seaven — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

**Pull request merged**